editor/components/block-settings-menu/block-duplicate-button.js

```diff
@@ -0,0 +1,64 @@
+/**
+ * External dependencies
+ */
+import { connect } from 'react-redux';
+import { flow, noop, last } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { IconButton, withContext } from '@wordpress/components';
+import { compose } from '@wordpress/element';
+import { cloneBlock } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { getBlocksByUid, getBlockIndex } from '../../store/selectors';
+import { insertBlocks } from '../../store/actions';
+
+export function BlockDuplicateButton( { onDuplicate, onClick = noop, isLocked, small = false } ) {
+	if ( isLocked ) {
+		return null;
+	}
+
+	const label = __( 'Duplicate' );
+
+	return (
+		<IconButton
+			className="editor-block-settings-menu__control"
+			onClick={ flow( onDuplicate, onClick ) }
+			icon="admin-page"
+			label={ small ? label : undefined }
+		>
+			{ ! small && label }
+		</IconButton>
+	);
+}
+
+export default compose(
+	connect(
+		( state, ownProps ) => ( {
+			blocks: getBlocksByUid( state, ownProps.uids ),
+			index: getBlockIndex( state, last( ownProps.uids ), ownProps.rootUID ),
+		} ),
+		{ insertBlocks },
+		( { blocks, index }, dispatchProps, { rootUID } ) => ( {
```

Comment:
I see why it's used here, but I've becoming increasingly cautious of `mergeProps` upon discovery in #5090 about excessive renders it can incur.

Reply:
Another option would be to do this inside the component. Do you think it's better?

Reply:
Better in performance. Worse in having it become less purely presentational.

🤷‍♂️

Reply:
Should be possible to simplify now with #5137